src/otk/annotation.py

```diff
@@ -0,0 +1,262 @@
+import copy
+import os
+import pathlib
+import sys
+from typing import TypeVar, Generic, Any, Optional
+
+
+class AnnotatedBase:
+    """ Implements a base class for annotated objects.
+    The annotations are basically a dict[str, any] of properties to attach to the class.
+    """
+
+    # base directory to calculate relative paths (easier to read)
+    _basedir = os.path.curdir
+
+    def __init__(self) -> None:
+        self._otk_annotations: dict[str, Any] = {}
+        if isinstance(self, list):
+            for idx, value in enumerate(self):
+                self[idx] = self.deep_convert(value)
+        elif isinstance(self, dict):
+            for key, value in self.items():
+                self[key] = self.deep_convert(value)
+
+    def get_annotation(self, key: str, default: Any = None) -> Any:
+        if key in self._otk_annotations:
+            return self._otk_annotations[key]
+        return default
+
+    def set_annotation(self, key: str, val: Any) -> Any:
+        self._otk_annotations[key] = val
+
+    def get_annotations(self) -> dict[str, Any]:
+        """Returns all annotations"""
+        return self._otk_annotations
+
+    def set_annotations(self, annotations: dict[str, Any]) -> None:
+        """ set all annotations """
+        self._otk_annotations = annotations
+
+    @classmethod
+    def deep_convert(cls, data: Any) -> Any:
+        """Recursively convert nested dicts and lists into AnnotatedDict and HiddenAttrList."""
```

Comment:
Right, I think this might be obsolete when also all tests use annotated objects when interacting with internals. I wanted to collect feedback first before changing all tests.